napari/components/layerlist.py

```diff
@@ -1,36 +1,50 @@
 from typing import Optional, List
 from ..layers import Layer
 from ..utils.naming import inc_name_count
-from ..utils.events.containers import TypedEventedList
+from ..utils.events import TypedEventedList
 
 
 class LayerList(TypedEventedList):
     """List-like layer collection with built-in reordering and callback hooks.
 
     Parameters
     ----------
-    data : iterable
+    iterable : iterable
         Iterable of napari.layer.Layer
+
+    Attributes
+    ----------
+    events : vispy.util.event.EmitterGroup
+        Event hooks:
+            * added(item, index): whenever an item is added
+            * removed(item): whenever an item is removed
+            * reordered(): whenever the list is reordered
     """
 
     def __init__(self, data=()):
         super().__init__(
-            basetype=Layer, data=data, lookup={str: lambda x: x.name},
-        )
-
-        self.events.inserted.connect(self._add)
-
-    def __newlike__(self, iterable):
-        return TypedEventedList(
-            basetype=self._basetype, data=iterable, lookup=self._lookup
+            data=data, basetype=Layer, lookup={str: lambda e: e.name},
         )
 
-    def _add(self, event):
-        """When a layer is added, set its name."""
-        layer = event.value
-        layer.name = self._coerce_name(layer.name, layer)
-        layer.events.name.connect(lambda e: self._update_name(e))
-        self.unselect_all(ignore=layer)
+    def _prepare_to_add_layer(self, layer: Layer):
+        if isinstance(layer, Layer):
```

Comment:
this happens before the type enforcement... so we still have to check if it's a layer.